### DISCUSSION TITLE
what's wrong with "ERROR:cairo_overlay.c:164:setup_gst_pipeline: assertion failed: (sink)"

### DISCUSSION THREAD

**Discussion by wuzz2zy** [svpcom/wfb-ng#117] [Questions]
I can't use it on CM4

**Comment by svpcom** [comment]
What gstreamer pipeline do you use?

**Comment by diphtong** [comment]
Hi,
i have the same error when running wifibroadcast_osd on Ubuntu 20.04
![Bildschirmfoto vom 2021-09-20 09-06-38](https://user-images.githubusercontent.com/91046964/133970955-64bb9bb0-fde4-47fd-b448-fe4d2e15f137.png)

**Reply by svpcom** [reply]
Try to replace `vaapisink` in cairo_overlay.c to `xvimagesink`. I use vaapisink by default becase xvimagesink doesn't support fullscreen output

**Reply by diphtong** [reply]
Hello,

thank you for your ready answer!

changing line 164 in cairo_overlay.c from     sink = gst_element_factory_make ("vaapisink", "sink"); to sink = gst_element_factory_make ("xvimagesink", "sink"); does not make a difference, same error.

**Reply by svpcom** [reply]
Try also replace ` g_object_set(sink, "sync", FALSE, "fullscreen", TRUE, NULL);` with ` g_object_set(sink, "sync", FALSE, NULL);`

**Reply by svpcom** [reply]
Fixed in https://github.com/svpcom/wifibroadcast_osd/commit/8e7cc0750d43eda6667441dc96565e1758c9640a

**Comment by diphtong** [comment]
hm, also no succss, same error

I have to run ./osd on the ground station, right?

**Reply by svpcom** [reply]
Yes. osd should be run on the ground station. Have you removed  `#define USE_VAAPI_OUTPUT` in cairo_overlay.c ?

**Reply by svpcom** [reply]
Try to run osd pipeline without osd binary:
```
gst-launch-1.0 udpsrc port=5600 caps='application/x-rtp, media=(string)video, clock-rate=(int)90000, encoding-name=(string)H264' ! rtph264depay ! avdec_h264 ! videoconvert ! cairooverlay ! videoconvert ! xvimagesink sync=false
```

**Reply by svpcom** [reply]
It should display video stream without osd text

**Reply by diphtong** [reply]
Hello,

yes, i commented out the #define USE_VAAPI_OUTPUT

When running the osd pipeline without osd binary, i got an output screen with the video stream i send to 127.0.0.1 on port 5600, so i think it works.

But when running the ./osd in the wifibroadcast_osd folder, i get alwasy the same error.

Thank you!

**Comment by diphtong** [comment]
Hello,

yes, i commented out the #define USE_VAAPI_OUTPUT

When i run the osd pipeline without osd binary, i get an video screen, which is showing the video i send to 127.0.0.1 on port 5600, so i think this work.

But when running ./osd i get still the same error.

**Reply by svpcom** [reply]
Have you got assertion errors related to sink in case of osd with xvimagesink?

**Reply by diphtong** [reply]
Good Morning!

when i run the osd (with commented out #define USE_VAAPI_OUTPUT), i got this error:

OSD_PORT=14551
RTP_PORT=5600
**
ERROR:cairo_overlay.c:164:setup_gst_pipeline: assertion failed: (sink)
Bail out! ERROR:cairo_overlay.c:164:setup_gst_pipeline: assertion failed: (sink)
Abgebrochen (Speicherabzug geschrieben)

when i have not commentetd out #define USE_VAAPI_OUTPUT, i get the same error.

I also get this error, when i do the modifications you mentioned.

Also running ./osd with sudo makes no difference.

Thank you very much for your help!

**Comment by diphtong** [comment]
Ok...i found the problem.

it works with xvimagesink, but...when making changes on cairo_overlay.c, i have to build it new with "make osd", when then running ./osd it works.

Is it also possible not to display the overlayed video in a gstreamer window, but streaming it to a local adress, to get the overlayed video for example in qgroundcontrol?

Thank you very much!

**Reply by svpcom** [reply]
IMHO in this case you don't need osd at all, just use qgroundcontrol with default settings (it will listen for RTP on udp 5600)

**Comment by diphtong** [comment]
yes i know, it work without osd.
But qgroundcontrol does not have such a nice osd video overlay. ;-)

**Reply by svpcom** [reply]
Why you need video both on osd and QGC?
You can send video stream to 224.0.0.1 multicast address instead of 127.0.0.1
```
[gs_video]
peer = 'connect://224.0.0.1:5600'
```
but this will not work for the same host (need to be osd and QGC on difference hosts).

I use QGC only for telemetry and drone control and osd on separate screen for video.

**Comment by diphtong** [comment]
Hm, a second screen would be an option...
i dont need both videos, actually only the osd video, which i wanted to stream directly to qgc...so maybe it is possible so that i can spare the second screen?

The mavlink port on osd is 14551. When i leave it, i dont get telemetry data for osd. When i change it in the main.c to port 14550, it works. But when i then run qgc, i have only telemetry data in qgc, not in osd anymore.
How do i have adjust it for working on both? Do i need a second mavlink-router stream to port 14551?

Thank you!

**Reply by svpcom** [reply]
I use:

```
[gs_mavlink]
mirror = 'connect://127.0.0.1:14551'
```
this mirror all incoming mavlink messages to udp:14551 (osd) while QGC use udp:14550
